File: cpp/src/Electron.cpp
#include "Electron.h"

#include "Photon.h"
#include "RandomMT.h"
#include "ResultStore.h"

Electron::Electron(double energy, double position,
                   Particle* parent, ResultStore *results) :
  Particle(energy, position, ELECTRON_RAD_LENGTH,
           ELECTRON_TERM_DECAY_LENGTH, parent, results)
{}

Electron::~Electron() {}

Particle::DecayType Electron::TerminalDecay() const {
  if(m_energy < ELECTRON_TERM_DECAY_ENERGY) {
    return Particle::IONIZATION;
  }

  return Particle::UNDECAYED;
}

Particle::DecayType Electron::Decay() {
  double brem_energy = RandomMT::Brem();

  m_child_1 = new Electron(m_energy * (1 - brem_energy),
                           m_position,
                           this,
                           m_results);

  m_child_2 = new Photon(m_energy * brem_energy,
                         m_position,
                         this,
                         m_results);

  return Particle::BREM;
}

void Electron::CountResult() {
  if(m_results) {
    m_results->IncrementNumChargedParticles();
  }
}

File: java/cosmicsim/src/main/java/com/ddamiani/cosmicsim/particle/Electron.java
package com.ddamiani.cosmicsim.particle;

import com.ddamiani.cosmicsim.RandomNumberHelper;
import com.ddamiani.cosmicsim.ResultStore;

/**
 * Class representing an electron - inherits from Particle.
 */
public final class Electron extends Particle {
    protected static final double ELECTRON_RAD_LENGTH = 0.28;
    protected static final double ELECTRON_TERM_DECAY_LENGTH = 1.0;
    protected static final double ELECTRON_TERM_DECAY_ENERGY = 100.0;

    public Electron(double energy, double position, Particle parent, ResultStore results) {
        super(energy, position, ELECTRON_RAD_LENGTH, ELECTRON_TERM_DECAY_LENGTH, parent, results);
    }

    protected final DecayType terminalDecay() {
        if (mEnergy < ELECTRON_TERM_DECAY_ENERGY) {
            return DecayType.IONIZATION;
        }

        return DecayType.UNDECAYED;
    }

    protected final DecayType decay() {
        double bremEnergy = RandomNumberHelper.brem();

        mChildOne = new Electron(mEnergy * (1 - bremEnergy),
                mPosition,
                this,
                mResults);

        mChildTwo = new Photon(mEnergy * bremEnergy,
                mPosition,
                this,
                mResults);

        return DecayType.BREM;
    }

    protected final void countResult() {
        if (mResults != null) {
            mResults.incrementNumChargedParticles();
        }
    }
}
